drivers/dma/dma_max32.c

```diff
@@ -0,0 +1,341 @@
+/*
+ * Copyright (c) 2023 Analog Devices, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/dma.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/irq.h>
+#include <zephyr/drivers/clock_control/adi_max32_clock_control.h>
+
+#include <wrap_max32_dma.h>
+
+#define DT_DRV_COMPAT adi_max32_dma
+
+LOG_MODULE_REGISTER(max32_dma, CONFIG_DMA_LOG_LEVEL);
+
+struct max32_dma_config {
+	mxc_dma_regs_t *regs;
+	const struct device *clock;
+	struct max32_perclk perclk;
+	uint8_t channels;
+	void (*irq_configure)(void);
+};
+
+struct max32_dma_data {
+	dma_callback_t callback;
+	void *cb_data;
+	uint32_t err_cb_dis;
+};
+
+static inline bool is_valid_dma_ch_prio(uint32_t ch_prio)
+{
+	/* mxc_dma_priority_t is limited to values 0-3 */
+	if (!(ch_prio >= 0 && ch_prio <= 3)) {
+		LOG_ERR("Invalid DMA priority - must be type mxc_dma_priority_t (0-3)");
+		return false;
+	}
+	return true;
+}
+
+static inline int get_dma_width(uint32_t width)
+{
+	switch (width) {
+	case 1:
+		return MXC_DMA_WIDTH_BYTE;
+	case 2:
+		return MXC_DMA_WIDTH_HALFWORD;
+	case 4:
+		return MXC_DMA_WIDTH_WORD;
+	default:
+		LOG_ERR("Invalid DMA width - must be byte (1), halfword (2) or word (4)");
+		return -EINVAL;
+	}
+}
+
+static inline int get_dma_addr_adj(uint16_t addr_adj)
+{
+	switch (addr_adj) {
+	case DMA_ADDR_ADJ_NO_CHANGE:
+		return 0;
+	case DMA_ADDR_ADJ_INCREMENT:
+		return 1;
+	default:
+		LOG_ERR("Invalid DMA address adjust - must be NO_CHANGE (0) or INCREMENT (1)");
+		return 0;
+	}
+}
+
+static inline int get_dma_ch_index(mxc_dma_regs_t *dma,  uint8_t ch)
+{
+	return (ch + MXC_DMA_GET_IDX(dma) * (MXC_DMA_CHANNELS / MXC_DMA_INSTANCES));
+}
+
+static int api_config(const struct device *dev, uint32_t channel,
+				   struct dma_config *config)
+{
+	int ret = 0;
+	const struct max32_dma_config *cfg = dev->config;
+	struct max32_dma_data *data = dev->data;
+	uint32_t ch;
+
+	if (channel >= cfg->channels) {
+		LOG_ERR("Invalid DMA channel - must be < %" PRIu32 " (%" PRIu32 ")", cfg->channels,
+			channel);
+		return -EINVAL;
+	}
+
+	ch = get_dma_ch_index(cfg->regs, channel);
+
+	/* DMA Channel Config */
+	mxc_dma_config_t mxc_dma_cfg;
+
+	mxc_dma_cfg.ch = ch;
+	mxc_dma_cfg.reqsel = config->dma_slot << ADI_MAX32_DMA_CFG_REQ_POS;
+	if (((get_dma_width(config->source_data_size)) < 0) ||
+	      ((get_dma_width(config->dest_data_size)) < 0)) {
+		return -EINVAL;
+	}
+	mxc_dma_cfg.srcwd = get_dma_width(config->source_data_size);
```

Comment:
would recommend renaming get_ functions to max32_dma_ prefixed functions to imply a conversion rather than an access type function, specific to the max32_dma

Reply:
Done.